src/atom_collection.rs

```diff
@@ -284,79 +264,72 @@ impl AtomCollection {
         }
     }
 
-    fn clean_up_consecutive(&self, atom: Atom, ignoring: &[Atom]) -> Vec<Atom> {
-        let filtered = self
-            .atoms
-            .split(|a| *a == atom)
-            .filter(|chain| !chain.is_empty())
-            .filter(|chain| !only(chain, ignoring));
-
-        Itertools::intersperse(filtered, &[atom.clone()])
-            .flatten()
-            .cloned()
-            .collect_vec()
-    }
-
-    fn trim_following(&self, delimiter: Atom, skip: Atom) -> Vec<Atom> {
-        let trimmed = self
-            .atoms
-            .split(|a| *a == delimiter)
-            .map(|slice| slice.iter().skip_while(|a| **a == skip).collect::<Vec<_>>());
-
-        Itertools::intersperse(trimmed, vec![&delimiter])
-            .flatten()
-            .cloned()
-            .collect_vec()
+    pub fn post_process(&mut self) {
+        let mut newvec = Vec::new();
+        post_process_internal(&mut newvec, &(self.atoms));
+        self.atoms = newvec;
     }
+}
 
-    fn put_before(&mut self, before: Atom, after: Atom, ignoring: &[Atom]) {
-        for i in 0..self.atoms.len() - 1 {
-            if self.atoms[i] == after {
-                for j in i + 1..self.atoms.len() {
-                    if self.atoms[j] != before
-                        && self.atoms[j] != after
-                        && !ignoring.contains(&self.atoms[j])
-                    {
-                        // stop looking
-                        break;
+fn post_process_internal(new_vec: &mut Vec<Atom>, old_vec: &Vec<Atom>) {
```

Comment:
I think we need a comment that describes the ordering that this function dictates.